### PR TITLE
gnat{6,9,10,11}: support building a cross-compiler

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Motivation for this change

This is somewhat working: `pkgsCross.aarch64-multiplatform.buildPackages.gnat` will now compile. However to achieve this it seems we need to disable building the runtime `libada` (see 9b107a9). As a result the GNAT cross compiler is not very useful.

Next thing to investigate is building `libada` as a standalone derivation.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/gcc/10/default.nix [hunk 1]
```diff
@@ -156,7 +156,9 @@ stdenv.mkDerivation ({
 
   depsBuildBuild = [ buildPackages.stdenv.cc ];
   nativeBuildInputs = [ texinfo which gettext ]
-    ++ (optional (perl != null) perl);
+    ++ (optional (perl != null) perl)
+    ++ (optional langAda gnatboot)
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gcc/10/default.nix:160]:
```suggestion
    ++ optional (perl != null) perl
    ++ optional langAda gnatboot
```

File: pkgs/development/compilers/gcc/11/default.nix [hunk 2]
```diff
@@ -161,7 +161,9 @@ stdenv.mkDerivation ({
 
   depsBuildBuild = [ buildPackages.stdenv.cc ];
   nativeBuildInputs = [ texinfo which gettext ]
-    ++ (optional (perl != null) perl);
+    ++ (optional (perl != null) perl)
+    ++ (optional langAda gnatboot)
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gcc/11/default.nix:165]:
```suggestion
    ++ optional (perl != null) perl
    ++ optional langAda gnatboot
```

File: pkgs/development/compilers/gcc/6/default.nix [hunk 3]
```diff
@@ -199,7 +199,9 @@ stdenv.mkDerivation ({
   nativeBuildInputs = [ texinfo which gettext ]
     ++ (optional (perl != null) perl)
     ++ (optional javaAwtGtk pkg-config)
-    ++ (optional (with stdenv.targetPlatform; isVc4 || isRedox) flex);
+    ++ (optional (with stdenv.targetPlatform; isVc4 || isRedox) flex)
+    ++ (optional langAda gnatboot)
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gcc/6/default.nix:203]:
```suggestion
    ++ optional (with stdenv.targetPlatform; isVc4 || isRedox) flex
    ++ optional langAda gnatboot
```

File: pkgs/development/compilers/gcc/9/default.nix [hunk 4]
```diff
@@ -170,7 +170,9 @@ stdenv.mkDerivation ({
 
   depsBuildBuild = [ buildPackages.stdenv.cc ];
   nativeBuildInputs = [ texinfo which gettext ]
-    ++ (optional (perl != null) perl);
+    ++ (optional (perl != null) perl)
+    ++ (optional langAda gnatboot)
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gcc/9/default.nix:174]:
```suggestion
    ++ optional (perl != null) perl
    ++ optional langAda gnatboot
```

File: pkgs/development/compilers/gnatboot/default.nix [hunk 5]
```diff
@@ -1,15 +1,19 @@
 { lib, stdenv, fetchurl }:
 
+if stdenv.hostPlatform != stdenv.targetPlatform
+then builtins.throw "gnatboot can't cross-compile"
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gnatboot/default.nix:4]:
```suggestion
then throw "gnatboot can't cross-compile"
```

File: pkgs/development/libraries/gcc/libada/default.nix [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  sourceRoot = "gcc-${gcc.version}/libada";
+
+  nativeBuildInputs = [
+    gnat
+  ];
+
+  preConfigure = ''
+    mkdir -p ../build
+    buildRoot="$(realpath ../build)"
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/gcc/libada/default.nix:24]: Missing meta.

Comment (sternenseemann) [reply pkgs/development/libraries/gcc/libada/default.nix]: was checked in by mistake

File: pkgs/development/compilers/gcc/10/default.nix (hunk 1 above)
Comment (sternenseemann) [reply pkgs/development/compilers/gcc/10/default.nix:160]: Won't start to reformat the gcc expression in this change, let's keep it consistent in itself.

File: pkgs/development/compilers/gnatboot/default.nix (hunk 5 above)
Comment (sternenseemann) [reply pkgs/development/compilers/gnatboot/default.nix:4]: Makes zero difference…

**Comment by Profpatsch** [comment]
Looks like a good progress, LGTM.

**Pull request merged**